main.c

```diff
@@ -50,5 +641,27 @@ void main() {
         // This puts the pins into the order R, G, R, G
         update_pwm(pwms >> 16, time_now, LED3);
         update_pwm(pwms >> 24, time_now, LED2);
+
+        // Handle GPIO direction set/clear
+        uint32_t val;
+        if ((val = SHARED_HACK->gpio_bank_01_dir_set)) {
+            GPIO_BANK_01->dir |= val;
+            SHARED_HACK->gpio_bank_01_dir_set = 0;
+        }
+        if ((val = SHARED_HACK->gpio_bank_01_dir_clr)) {
+            GPIO_BANK_01->dir &= ~val;
+            SHARED_HACK->gpio_bank_01_dir_clr = 0;
+        }
+
+        // Handle I2C
+        if ((*TIMER2_INTCTLSTAT) & TIMER_INTCTLSTAT_PRDINTSTAT12) {
+            *TIMER2_INTCTLSTAT = TIMER_INTCTLSTAT_PRDINTSTAT12;
+
+            // This code runs at 20 kHz, which is 2x the I2C clock rate
+            handle_i2c(&SHARED->i2c[0], &i2c_states[0]);
```

Comment:
Just curious why `SHARED_HACK` doesn't work here. Because of the additional array indexing? Or are we at an offset beyond what the optimized instruction can handle?

Reply:
The special `lbco`/`sbco` opcodes only work when you actually load/store through the constant table address. They can't be used to compute an effective address (no equivalent of x86 opcode `LEA`). The only way to get this address is to actually load it as an immediate, in which case we need the real value.